clippy_lints/src/path_from_format.rs

```diff
@@ -0,0 +1,136 @@
+            then {
+                let format_string_parts = format_args.format_string_parts;
+                let format_value_args = format_args.value_args;
+                let mut string_parts: Vec<&str> = format_string_parts.iter().map(rustc_span::Symbol::as_str).collect();
+                string_parts.push("");
+                let mut applicability = Applicability::MachineApplicable;
+                let real_vars: Vec<Sugg<'_>> = format_value_args.iter().map(|x| Sugg::hir_with_applicability(cx, x, "..", &mut applicability)).collect();
+                let order_of_real_vars: Vec<usize> = format_args.formatters.iter().map(|(x, _)| *x).collect();
+                let mut sugg = String::new();
+                for n in 0..real_vars.len() {
+                    if !(
+                            string_parts[n].is_empty()
+                            || string_parts[n].ends_with('/')
+                            || string_parts[n].ends_with('\\')
+                        )
+                        || !(
+                            string_parts[n+1].is_empty()
+                            || string_parts[n+1].starts_with('/')
+                            || string_parts[n+1].starts_with('\\')
+                        ){
+                        span_lint_and_note(
+                            cx,
+                            PATH_FROM_FORMAT,
+                            expr.span,
+                            "`format!(..)` used to form `PathBuf`",
+                            None,
+                            "if it fits your use case, you may want to consider using `Path::new()` and `.join()` to make it OS-agnostic and improve code readability",
+                        );
+                        return;
+                    }
+                    if n == 0 {
+                        if string_parts[0].is_empty() {
+                            sugg = format!("Path::new({})", real_vars[order_of_real_vars[0]]);
+                        }
+                        else {
+                            push_comps(&mut sugg, Path::new(string_parts[0]));
+                            let _ = write!(sugg, ".join({})", real_vars[order_of_real_vars[0]]);
+                        }
+                    }
+                    else if string_parts[n].is_empty() {
+                            sugg = format!("{sugg}.join({})", real_vars[order_of_real_vars[n]]);
+                    }
+                    else {
+                        let mut string = String::from(string_parts[n]);
+                        if string.starts_with('/') || string.starts_with('\\') {
+                            string.remove(0);
+                        }
+                        push_comps(&mut sugg, Path::new(&string));
+                        let _ = write!(sugg, ".join({})", real_vars[order_of_real_vars[n]]);
+                    }
+                }
+                if !string_parts[real_vars.len()].is_empty() {
```

Comment:
I'm assuming the empty string push in Line 58 is for this scenario, since `string_parts.len` is either `real_vars.len` or `real_vars.len + 1`. the empty string is to ensure that `string_parts.len > real_vars.len`. 

If that's the case, then we can just check if `real_vars.len < string_parts.len`. Then add the last `string_parts` to the suggestion..